addons/account/models/account_move.py

```diff
@@ -404,19 +414,40 @@ def _get_accounting_date(self, invoice_date, has_tax):
                     return max(invoice_date, today)
         return invoice_date
 
+    def _get_closest_violated_lock_date_and_type(self, invoice_date, has_tax):
+        tax_lock_date = self.company_id.tax_lock_date
+        user_lock_date = self.company_id._get_user_fiscal_lock_date()
+        tax_lock_violated = invoice_date and tax_lock_date and has_tax and invoice_date <= tax_lock_date
+        user_lock_violated = invoice_date and user_lock_date and invoice_date <= user_lock_date
+        closest_lock_date = max(tax_lock_violated and tax_lock_date or date.min, user_lock_violated and user_lock_date or date.min)
+        lock_type = False
+        if tax_lock_violated and closest_lock_date == tax_lock_date:
+            lock_type = 'tax'
+        elif user_lock_violated and closest_lock_date == user_lock_date:
+            lock_type = 'user'
+        return closest_lock_date, lock_type
```

Comment:
What about returning the list of violated locks as a list of tuple `(lock_date, lock_type)`, so that the error message outputs all the issues at the same time?


```suggestion
        """Some docstring."""
        locks = []
        user_lock_date = self.company_id._get_user_fiscal_lock_date()
        if invoice_date and user_lock_date and invoice_date <= user_lock_date:
            locks.append((user_lock_date, 'user'))
        tax_lock_date = self.company_id.tax_lock_date
        if invoice_date and tax_lock_date and has_tax and invoice_date <= tax_lock_date:
            locks.append((tax_lock_date, 'tax'))
        return locks
```

You could return an ordered version of `locks`, or get the maximum of the lock dates when setting it